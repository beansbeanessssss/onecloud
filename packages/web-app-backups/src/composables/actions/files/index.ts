export * from './useFileActionsCreateQuicklink'
export * from './useFileActionsPaste'
export * from './useFileActionsShowDetails'
export * from './useFileActionsShowShares'
export * from './useFileActionsAcceptShare'
export * from './useFileActionsCopy'
export * from './useFileActionsDeclineShare'
export * from './useFileActionsDelete'
export * from './useFileActionsDownloadArchive'
export * from './useFileActionsDownloadFile'
export * from './useFileActionsEmptyTrashBin'
export * from './useFileActionsMove'
export * from './useFileActionsRestore'
export * from './useFileActionsRestore'
export * from './useFileActionsRename'
export * from './useFileActionsSetImage'
export * from './useFileActionsShowEditTags'
export * from './useFileActionsNavigate'
export * from './useFileActionsFavorite'
export * from './useFileActionsCreateSpaceFromResource'
