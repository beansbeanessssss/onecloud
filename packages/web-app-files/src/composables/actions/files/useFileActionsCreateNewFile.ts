import { Resource, extractNameWithoutExtension } from 'web-client/src/helpers'
import { Store } from 'vuex'
import { computed, ref, unref } from 'vue'
import { useClientService, useRequest, useRouter, useStore } from 'web-pkg/src/composables'
import { FileAction } from 'web-pkg/src/composables/actions'
import { useGettext } from 'vue3-gettext'
import { resolveFileNameDuplicate } from 'web-app-files/src/helpers/resource'
import { join } from 'path'
import { WebDAV } from 'web-client/src/webdav'
import { isLocationSpacesActive } from 'web-app-files/src/router'
import { getIndicators } from 'web-app-files/src/helpers/statusIndicators'
import { EDITOR_MODE_CREATE, useFileActions } from './useFileActions'
import { urlJoin } from 'web-client/src/utils'
import { configurationManager } from 'web-pkg/src'
import { stringify } from 'qs'

export const useFileActionsCreateNewFile = ({
  store,
  openAction,
  addAppProviderFile,
  extension
}: {
  store?: Store<any>
  openAction?: any
  addAppProviderFile?: boolean
  extension?: string
} = {}) => {
  store = store || useStore()
  const router = useRouter()
  const { $gettext } = useGettext()
  const { makeRequest } = useRequest()

  const { openEditor, triggerDefaultAction } = useFileActions()
  const clientService = useClientService()
  const currentFolder = computed((): Resource => store.getters['Files/currentFolder'])
  const files = computed((): Array<Resource> => store.getters['Files/files'])
  const ancestorMetaData = computed(() => store.getters['Files/ancestorMetaData'])
  const areFileExtensionsShown = computed(
    (): boolean => store.getters['Files/areFileExtensionsShown']
  )
  const storageId = computed(() => unref(currentFolder).storageId)
  const currentSpace = computed(() =>
    store.getters['runtime/spaces/spaces'].find((space) => space.id === unref(storageId))
  )
  const newFileAction = ref(null)

  const capabilities = computed(() => store.getters['capabilities'])

  const checkNewFileName = (fileName) => {
    if (fileName === '') {
      return $gettext('File name cannot be empty')
    }

    if (/[/]/.test(fileName)) {
      return $gettext('File name cannot contain "/"')
    }

    if (fileName === '.') {
      return $gettext('File name cannot be equal to "."')
    }

    if (fileName === '..') {
      return $gettext('File name cannot be equal to ".."')
    }

    if (/\s+$/.test(fileName)) {
      return $gettext('File name cannot end with whitespace')
    }

    const exists = unref(files).find((file) => file.name === fileName)

    if (exists) {
      const translated = $gettext('%{name} already exists')
      return $gettext(translated, { name: fileName }, true)
    }

    return null
  }

  const addAppProviderFileFunc = async (fileName) => {
    // FIXME: this belongs in web-app-external, but the app provider handles file creation differently than other editor extensions. Needs more refactoring.
    if (fileName === '') {
      return
    }
    try {
      const baseUrl = urlJoin(
        configurationManager.serverUrl,
        unref(capabilities).files.app_providers[0].new_url
      )
      const query = stringify({
        parent_container_id: unref(currentFolder).fileId,
        filename: fileName
      })
      const url = `${baseUrl}?${query}`
      const response = await makeRequest('POST', url)
      if (response.status !== 200) {
        throw new Error(`An error has occurred: ${response.status}`)
      }
      const path = join(unref(currentFolder).path, fileName) || ''
      const resource = await (clientService.webdav as WebDAV).getFileInfo(unref(currentSpace), {
        path
      })
      if (unref(loadIndicatorsForNewFile)) {
        resource.indicators = getIndicators({ resource, ancestorMetaData: unref(ancestorMetaData) })
      }
      triggerDefaultAction({ space: unref(currentSpace), resources: [resource] })
      store.commit('Files/UPSERT_RESOURCE', resource)
      store.dispatch('hideModal')
      store.dispatch('showMessage', {
        title: $gettext('"%{fileName}" was created successfully', { fileName })
      })
    } catch (error) {
      console.error(error)
      store.dispatch('showMessage', {
        title: $gettext('Failed to create file'),
        status: 'danger'
      })
    }
  }

  const loadIndicatorsForNewFile = computed(() => {
    return (
      isLocationSpacesActive(router, 'files-spaces-projects') &&
      unref(currentSpace).driveType !== 'share'
    )
  })

  const addNewFile = async (fileName) => {
    if (fileName === '') {
      return
    }

    try {
      const path = join(unref(currentFolder).path, fileName)
      const resource = await (clientService.webdav as WebDAV).putFileContents(unref(currentSpace), {
        path
      })

      if (loadIndicatorsForNewFile) {
        resource.indicators = getIndicators({ resource, ancestorMetaData: unref(ancestorMetaData) })
      }

      store.commit('Files/UPSERT_RESOURCE', resource)

      if (unref(newFileAction)) {
        openEditor(
          unref(newFileAction),
          unref(currentSpace).getDriveAliasAndItem(resource),
          resource.webDavPath,
          resource.fileId,
          EDITOR_MODE_CREATE
        )
        store.dispatch('hideModal')

        return
      }

      store.dispatch('hideModal')
      store.dispatch('showMessage', {
        title: $gettext('"%{fileName}" was created successfully', { fileName })
      })
    } catch (error) {
      console.error(error)
      store.dispatch('showMessage', {
        title: $gettext('Failed to create file'),
        status: 'danger'
      })
    }
  }

  const handler = () => {
    const checkInputValue = (value) => {
      store.dispatch(
        'setModalInputErrorMessage',
        checkNewFileName(areFileExtensionsShown ? value : `${value}.${extension}`)
      )
    }
    let defaultName = $gettext('New file') + `.${extension}`

    if (unref(files).some((f) => f.name === defaultName)) {
      defaultName = resolveFileNameDuplicate(defaultName, extension, unref(files))
    }

    if (!areFileExtensionsShown) {
      defaultName = extractNameWithoutExtension({ name: defaultName, extension } as any)
    }

    // Sets action to be executed after creation of the file
    newFileAction.value = openAction

    const inputSelectionRange = !areFileExtensionsShown
      ? null
      : [0, defaultName.length - (extension.length + 1)]

    const modal = {
      variation: 'passive',
      title: $gettext('Create a new file'),
      cancelText: $gettext('Cancel'),
      confirmText: $gettext('Create'),
      hasInput: true,
      inputValue: defaultName,
      inputLabel: $gettext('File name'),
      inputError: checkNewFileName(
        areFileExtensionsShown ? defaultName : `${defaultName}.${extension}`
      ),
      inputSelectionRange,
      onCancel: () => store.dispatch('hideModal'),
      onConfirm: addAppProviderFile
        ? addAppProviderFileFunc
        : (fileName) => {
            if (!areFileExtensionsShown) {
              fileName = `${fileName}.${extension}`
            }
            addNewFile(fileName)
          },
      onInput: checkInputValue
    }

    store.dispatch('createModal', modal)
  }

  const actions = computed((): FileAction[] => {
    return [
      {
        name: 'create-new-file',
        icon: 'add',
        handler,
        label: () => {
          return $gettext('Create new File')
        },
        isEnabled: ({ resources }) => {
          return true
        },
        canBeDefault: true,
        componentType: 'button',
        class: 'oc-files-actions-create-new-file'
      }
    ]
  })

  return {
    actions
  }
}
