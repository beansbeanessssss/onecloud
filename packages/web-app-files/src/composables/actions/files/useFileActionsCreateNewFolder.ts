import { Resource, extractNameWithoutExtension } from 'web-client/src/helpers'
import { Store } from 'vuex'
import { computed, unref } from 'vue'
import { useClientService, useRouter, useStore } from 'web-pkg/src/composables'
import { FileAction, FileActionOptions } from 'web-pkg/src/composables/actions'
import { useGettext } from 'vue3-gettext'
import { resolveFileNameDuplicate } from 'web-app-files/src/helpers/resource'
import { join } from 'path'
import { WebDAV } from 'web-client/src/webdav'
import { isLocationSpacesActive } from 'web-app-files/src/router'
import { getIndicators } from 'web-app-files/src/helpers/statusIndicators'

export const useFileActionsCreateNewFolder = ({ store }: { store?: Store<any> } = {}) => {
  store = store || useStore()
  const router = useRouter()
  const { $gettext } = useGettext()

  const clientService = useClientService()
  const currentFolder = computed((): Resource => store.getters['Files/currentFolder'])
  const files = computed((): Array<Resource> => store.getters['Files/files'])
  const ancestorMetaData = computed(() => store.getters['Files/ancestorMetaData'])
  const areFileExtensionsShown = computed(
    (): boolean => store.getters['Files/areFileExtensionsShown']
  )

  const storageId = unref(currentFolder).storageId
  const currentSpace = store.getters['runtime/spaces/spaces'].find(
    (space) => space.id === storageId
  ) //TODO is this the way to replace this.space ( ͡° ͜ʖ ͡°)?

  const checkNewFolderName = (folderName) => {
    if (folderName.trim() === '') {
      return $gettext('Folder name cannot be empty')
    }

    if (/[/]/.test(folderName)) {
      return $gettext('Folder name cannot contain "/"')
    }

    if (folderName === '.') {
      return $gettext('Folder name cannot be equal to "."')
    }

    if (folderName === '..') {
      return $gettext('Folder name cannot be equal to ".."')
    }

    const exists = unref(files).find((file) => file.name === folderName)

    if (exists) {
      const translated = $gettext('%{name} already exists')
      return $gettext(translated, { name: folderName }, true)
    }

    return null
  }

  const checkNewFileName = (fileName) => {
    if (fileName === '') {
      return $gettext('File name cannot be empty')
    }

    if (/[/]/.test(fileName)) {
      return $gettext('File name cannot contain "/"')
    }

    if (fileName === '.') {
      return $gettext('File name cannot be equal to "."')
    }

    if (fileName === '..') {
      return $gettext('File name cannot be equal to ".."')
    }

    if (/\s+$/.test(fileName)) {
      return $gettext('File name cannot end with whitespace')
    }

    const exists = unref(files).find((file) => file.name === fileName)

    if (exists) {
      const translated = $gettext('%{name} already exists')
      return $gettext(translated, { name: fileName }, true)
    }

    return null
  }

  const addAppProviderFileFunc = async (fileName) => {}

  const loadIndicatorsForNewFile = computed(() => {
    return (
      isLocationSpacesActive(router, 'files-spaces-projects') && currentSpace.driveType !== 'share'
    )
  })

  const addNewFolder = async (folderName) => {
    folderName = folderName.trimEnd()

    try {
      const path = join(unref(currentFolder).path, folderName) // TODO: Why was this.item used here isn't it just the currentFolderPath?
      const resource = await (clientService.webdav as WebDAV).createFolder(currentSpace, {
        path
      })

      if (unref(loadIndicatorsForNewFile)) {
        resource.indicators = getIndicators({ resource, ancestorMetaData: unref(ancestorMetaData) })
      }

      store.commit('Files/UPSERT_RESOURCE', resource)
      store.dispatch('hideModal')

      store.dispatch('showMessage', {
        title: $gettext('"%{folderName}" was created successfully', { folderName })
      })
    } catch (error) {
      console.error(error)
      store.dispatch('showMessage', {
        title: $gettext('Failed to create folder'),
        status: 'danger'
      })
    }
  }

  const handler = ({ space, resources }: FileActionOptions) => {
    const checkInputValue = (value) => {
      /*this.setModalInputErrorMessage(
        isFolder
          ? this.checkNewFolderName(value)
          : this.checkNewFileName(this.areFileExtensionsShown ? value : `${value}.${ext}`) //TODO
      )*/
    }
    const addAppProviderFile = false
    const isFolder = true
    const ext = ''
    let defaultName = isFolder ? $gettext('New folder') : $gettext('New file') + `.${ext}`

    if (unref(files).some((f) => f.name === defaultName)) {
      defaultName = resolveFileNameDuplicate(defaultName, isFolder ? '' : ext, unref(files))
    }

    if (!areFileExtensionsShown) {
      defaultName = extractNameWithoutExtension({ name: defaultName, extension: ext } as any)
    }

    // Sets action to be executed after creation of the file
    if (!isFolder) {
      //this.newFileAction = openAction // TODO
    }

    const inputSelectionRange =
      isFolder || !areFileExtensionsShown ? null : [0, defaultName.length - (ext.length + 1)]

    const modal = {
      variation: 'passive',
      title: isFolder ? $gettext('Create a new folder') : $gettext('Create a new file'),
      cancelText: $gettext('Cancel'),
      confirmText: $gettext('Create'),
      hasInput: true,
      inputValue: defaultName,
      inputLabel: isFolder ? $gettext('Folder name') : $gettext('File name'),
      inputError: isFolder
        ? checkNewFolderName(defaultName)
        : checkNewFileName(areFileExtensionsShown ? defaultName : `${defaultName}.${ext}`),
      inputSelectionRange,
      onCancel: () => store.dispatch('hideModal'),
      onConfirm: isFolder
        ? addNewFolder
        : addAppProviderFile
        ? addAppProviderFileFunc
        : (fileName) => {
            /*if (!this.areFileExtensionsShown) {
              fileName = `${fileName}.${ext}`
            }
            this.addNewFile(fileName)*/
            //TODO
          },
      onInput: checkInputValue
    }

    store.dispatch('createModal', modal)
  }

  const actions = computed((): FileAction[] => {
    return [
      {
        name: 'create-folder',
        icon: 'add',
        handler,
        label: () => {
          return $gettext('Create new Folder')
        },
        isEnabled: ({ resources }) => {
          return true
        },
        canBeDefault: true,
        componentType: 'button',
        class: 'oc-files-actions-download-archive-trigger'
      }
    ]
  })

  return {
    actions
  }
}
