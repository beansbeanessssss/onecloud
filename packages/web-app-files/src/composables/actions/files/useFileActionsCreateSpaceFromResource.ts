import { Store } from 'vuex'
import { computed, unref } from 'vue'
import { useGettext } from 'vue3-gettext'
import { FileAction, FileActionOptions } from 'web-pkg/src/composables/actions'
import { useAbility, useClientService, useLoadingService, useRouter } from 'web-pkg/src/composables'
import { isPersonalSpaceResource } from 'web-client/src/helpers'
import { isLocationSpacesActive } from 'web-app-files/src/router'
import { useCreateSpace } from 'web-app-files/src/composables/spaces'
import { useSpaceHelpers } from 'web-pkg/src/composables/spaces'
import PQueue from 'p-queue'

export const useFileActionsCreateSpaceFromResource = ({ store }: { store?: Store<any> } = {}) => {
  const { can } = useAbility()
  const { $gettext, $ngettext } = useGettext()
  const loadingService = useLoadingService()
  const { createSpace } = useCreateSpace()
  const { checkSpaceNameModalInput } = useSpaceHelpers()
  const { webdav } = useClientService()
  const router = useRouter()
  const hasCreatePermission = computed(() => can('create-all', 'Space'))

  const confirmAction = async ({ spaceName, resources, space }) => {
    store.dispatch('hideModal')
    const queue = new PQueue({ concurrency: 4 })
    const copyOps = []

    try {
      const createdSpace = await createSpace(spaceName)
      store.commit('runtime/spaces/UPSERT_SPACE', createdSpace)

      if (resources.length === 1 && resources[0].isFolder) {
        //If a single folder is selected we copy it's content to the Space's root folder
        resources = (await webdav.listFiles(space, { path: resources[0].path })).children
      }

      for (const resource of resources) {
        copyOps.push(
          queue.add(() => webdav.copyFiles(space, resource, createdSpace, { path: resource.name }))
        )
      }

      await Promise.all(copyOps)
      store.dispatch('Files/resetFileSelection')
      store.dispatch('showMessage', {
        title: $gettext('Space was created successfully')
      })
    } catch (error) {
      console.error(error)
      store.dispatch('showMessage', {
        title: $gettext('Creating space failedâ€¦'),
        status: 'danger'
      })
    }
  }
  const handler = ({ resources, space }: FileActionOptions) => {
    const modal = {
      variation: 'passive',
      title: $ngettext(
        'Create Space from "%{resourceName}"',
        'Create Space from selection',
        resources.length,
        {
          resourceName: resources[0].name
        }
      ),
      message: $ngettext(
        'Create Space with the content of "%{resourceName}". The content will be copied.',
        'Create Space with the selected files. The content will be copied.',
        resources.length,
        {
          resourceName: resources[0].name
        }
      ),
      cancelText: $gettext('Cancel'),
      confirmText: $gettext('Create'),
      hasInput: true,
      inputLabel: $gettext('Space name'),
      onInput: checkSpaceNameModalInput,
      onCancel: () => store.dispatch('hideModal'),
      onConfirm: (spaceName) =>
        loadingService.addTask(() => confirmAction({ spaceName, space, resources }))
    }

    store.dispatch('createModal', modal)
  }

  const actions = computed((): FileAction[] => {
    return [
      {
        name: 'create-space-from-resource',
        icon: 'layout-grid',
        handler,
        label: () => {
          return $gettext('Create Space from selection')
        },
        isEnabled: ({ resources, space }) => {
          if (!resources.length) {
            return false
          }

          if (!unref(hasCreatePermission)) {
            return false
          }

          if (
            !isLocationSpacesActive(router, 'files-spaces-generic') ||
            !isPersonalSpaceResource(space)
          ) {
            return false
          }

          return true
        },
        componentType: 'button',
        class: 'oc-files-actions-create-space-from-resource-trigger'
      }
    ]
  })

  return {
    actions
  }
}
