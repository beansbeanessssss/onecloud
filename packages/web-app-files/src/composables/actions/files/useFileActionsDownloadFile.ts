import {
  isLocationCommonActive,
  isLocationPublicActive,
  isLocationSharesActive,
  isLocationSpacesActive
} from '../../../router'
import { useIsFilesAppActive } from '../helpers/useIsFilesAppActive'
import { useRouter } from 'web-pkg/src/composables'
import { Action, useIsSearchActive } from 'web-pkg/src/composables/actions'
import { computed, unref } from 'vue'
import { useGettext } from 'vue3-gettext'
import { useDownloadFile as _useDownloadFile } from 'web-pkg/src/composables/download/useDownloadFile'

export const useFileActionsDownloadFile = () => {
  const router = useRouter()
  const { $gettext } = useGettext()
  const isFilesAppActive = useIsFilesAppActive()
  const isSearchActive = useIsSearchActive()
  const { downloadFile } = _useDownloadFile()
  const handler = ({ resources }) => {
    downloadFile(resources[0])
  }

  const actions = computed((): Action[] => [
    {
      name: 'download-file',
      icon: 'file-download',
      handler,
      label: () => {
        return $gettext('Download')
      },
      isEnabled: ({ resources }) => {
        if (
          unref(isFilesAppActive) &&
          !unref(isSearchActive) &&
          !isLocationSpacesActive(router, 'files-spaces-generic') &&
          !isLocationPublicActive(router, 'files-public-link') &&
          !isLocationCommonActive(router, 'files-common-favorites') &&
          !isLocationCommonActive(router, 'files-common-search') &&
          !isLocationSharesActive(router, 'files-shares-with-me') &&
          !isLocationSharesActive(router, 'files-shares-with-others') &&
          !isLocationSharesActive(router, 'files-shares-via-link')
        ) {
          return false
        }
        if (resources.length !== 1) {
          return false
        }
        if (resources[0].isFolder) {
          return false
        }
        return resources[0].canDownload()
      },
      canBeDefault: true,
      componentType: 'button',
      class: 'oc-files-actions-download-file-trigger'
    }
  ])

  return {
    actions
  }
}
