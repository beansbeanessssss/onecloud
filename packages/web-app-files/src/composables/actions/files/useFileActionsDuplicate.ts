import {
  isLocationCommonActive,
  isLocationPublicActive,
  isLocationSpacesActive
} from '../../../router'
import { Store } from 'vuex'
import { computed, unref } from 'vue'
import {
  useClientService,
  useGetMatchingSpace,
  useLoadingService,
  useRouter,
  useStore
} from 'web-pkg/src/composables'
import { FileAction, FileActionOptions } from 'web-pkg/src/composables/actions'
import { isProjectSpaceResource } from 'web-client/src/helpers'
import { useGettext } from 'vue3-gettext'

export const useFileActionsDuplicate = ({ store }: { store?: Store<any> } = {}) => {
  store = store || useStore()
  const router = useRouter()
  const clientService = useClientService()
  const loadingService = useLoadingService()
  const { getMatchingSpace } = useGetMatchingSpace()
  const { $gettext, $pgettext, interpolate: $gettextInterpolate, $ngettext } = useGettext()

  const language = useGettext()

  const isMacOs = computed(() => {
    return window.navigator.platform.match('Mac')
  })

  const duplicateShortcutString = computed(() => {
    if (unref(isMacOs)) {
      return $pgettext('Keyboard shortcut for macOS for copying files', 'âŒ˜ + C + V')
    }
    return $pgettext('Keyboard shortcut for non-macOS systems for copying files', 'Ctrl + C + V')
  })

  const handler = async ({ space: targetSpace, resources }: FileActionOptions) => {
    try {
      const resourceSpaceMapping: Record<string, { space: SpaceResource; resources: Resource[] }> =
        resources.reduce((acc, resource) => {
          if (resource.storageId in acc) {
            acc[resource.storageId].resources.push(resource)
            return acc
          }

          const matchingSpace = getMatchingSpace(resource)

          if (!(matchingSpace.id in acc)) {
            acc[matchingSpace.id] = { space: matchingSpace, resources: [] }
          }

          acc[matchingSpace.id].resources.push(resource)
          return acc
        }, {})

      const promises = Object.values(resourceSpaceMapping).map(
        ({ space: sourceSpace, resources: resourcesToCopy }) => {
          return loadingService.addTask(() => {
            return store.dispatch('Files/duplicateSelectedFiles', {
              targetSpace,
              sourceSpace: sourceSpace,
              resources: resourcesToCopy,
              clientService,
              loadingService,
              createModal: (...args) => store.dispatch('createModal', ...args),
              hideModal: (...args) => store.dispatch('hideModal', ...args),
              showMessage: (...args) => store.dispatch('showMessage', ...args),
              $gettext,
              $gettextInterpolate,
              $ngettext
            })
          })
        }
      )
      await Promise.all(promises)
    } catch (error) {
      console.error(error)
    }
  }

  const actions = computed((): FileAction[] => {
    return [
      {
        name: 'duplicate',
        icon: 'file-copy-2',
        handler,
        shortcut: unref(duplicateShortcutString),
        label: ({ resources }) => {
          const duplicateLabel = $pgettext(
            'Action in the files list row to initiate duplicating resources',
            'Duplicate'
          )

          if (isLocationCommonActive(router, 'files-common-search') && resources.length > 1) {
            const copyableResourcesCount = resources.filter(
              (r) => !isProjectSpaceResource(r)
            ).length
            return `${duplicateLabel} (${copyableResourcesCount.toString()})`
          }

          return duplicateLabel
        },
        isEnabled: ({ resources }) => {
          if (
            !isLocationSpacesActive(router, 'files-spaces-generic') &&
            !isLocationPublicActive(router, 'files-public-link') &&
            !isLocationCommonActive(router, 'files-common-favorites')
          ) {
            return false
          }
          if (isLocationSpacesActive(router, 'files-spaces-projects')) {
            return false
          }
          if (resources.length === 0) {
            return false
          }

          if (isLocationPublicActive(router, 'files-public-link')) {
            return store.getters['Files/currentFolder'].canCreate()
          }

          if (
            isLocationCommonActive(router, 'files-common-search') &&
            resources.every((r) => isProjectSpaceResource(r))
          ) {
            return false
          }

          // duplicate can't be restricted in authenticated context, because
          // a user always has their home dir with write access
          return true
        },
        componentType: 'button',
        class: 'oc-files-actions-duplicate-trigger'
      }
    ]
  })

  return {
    actions
  }
}
