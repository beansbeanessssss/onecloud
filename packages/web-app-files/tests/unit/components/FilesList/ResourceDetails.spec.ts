import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  defaultStoreMockOptions,
  mount,
  RouteLocation
} from 'web-test-helpers'
import ResourceDetails from '../../../../src/components/FilesList/ResourceDetails.vue'
import { mock } from 'jest-mock-extended'
import { useFileActions } from 'web-app-files/src/composables/actions/files/useFileActions'
import { SpaceResource } from 'web-client'
import { useRouteQuery } from 'web-pkg'
import { ref } from 'vue'

jest.mock('web-app-files/src/composables/actions/files/useFileActions')
jest.mock('web-pkg/src/composables/router', () => ({
  ...jest.requireActual('web-pkg/src/composables/router'),
  useRouteQuery: jest.fn()
}))

describe('ResourceDetails component', () => {
  jest.mocked(useFileActions).mockImplementation(() =>
    mock<ReturnType<typeof useFileActions>>({
      getDefaultEditorAction: () => ({ handler: jest.fn() } as any)
    })
  )

  it('renders resource details correctly', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('open with default actions', () => {
    it("doesn't open default action if query param 'openWithDefaultApp' isn't set true", () => {
      const { wrapper } = getWrapper()
      expect(wrapper.vm.defaultEditorAction.handler).not.toHaveBeenCalled()
    })
    it("opens default action if query param 'openWithDefaultApp' is set true", () => {
      jest.mocked(useRouteQuery).mockImplementationOnce(() => ref('true'))
      const { wrapper } = getWrapper()
      expect(wrapper.vm.defaultEditorAction.handler).toHaveBeenCalled()
    })
  })

  function getWrapper() {
    const mocks = {
      ...defaultComponentMocks({
        currentRoute: mock<RouteLocation>({
          name: 'files-public-link',
          query: { openWithDefaultAppQuery: 'true' }
        })
      })
    }
    const storeOptions = { ...defaultStoreMockOptions }
    const store = createStore(storeOptions)

    const file = {
      id: 0,
      name: 'image.jpg',
      size: 24064,
      mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
      mimeType: 'image/jpg',
      owner: [
        {
          username: 'admin'
        }
      ]
    }
    const space = mock<SpaceResource>()

    return {
      mocks,
      wrapper: mount(ResourceDetails, {
        props: {
          space,
          singleResource: file
        },
        global: {
          mocks,
          plugins: [...defaultPlugins(), store],
          provide: {
            space,
            resource: file
          }
        }
      })
    }
  }
})
