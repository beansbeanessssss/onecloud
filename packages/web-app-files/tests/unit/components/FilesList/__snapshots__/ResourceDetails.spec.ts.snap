// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceDetails component renders resource details correctly 1`] = `
<div class="resource-details oc-flex oc-flex-column oc-flex-middle">
  <div class="oc-width-1-3@l oc-width-1-2@m oc-width-3-4">
    <div class="file_info oc-flex oc-flex-between oc-p-s">
      <div class="oc-flex oc-flex-middle">
        <span class="file_info__icon oc-mr-s">
          <span class="oc-icon oc-icon-l oc-icon-passive oc-resource-icon oc-resource-icon-folder">
            <!---->
          </span>
          <!--v-if-->
        </span>
        <div class="file_info__body oc-text-overflow">
          <h3 class="oc-font-semibold">
            <span class="oc-resource-name oc-display-inline-block" data-test-resource-name="image.jpg" title="image.jpg">
              <span class="oc-resource-basename oc-text-break">image.jpg</span>
              <!--v-if-->
            </span>
          </h3>
        </div>
      </div>
      <!--v-if-->
    </div>
    <div class="oc-mb" id="oc-file-details-sidebar">
      <div>
        <div class="details-icon-wrapper oc-width-1-1 oc-flex oc-flex-middle oc-flex-center oc-mb">
          <span class="details-icon">
            <span class="oc-icon oc-icon-xxxl oc-icon-passive oc-resource-icon oc-resource-icon-folder">
              <!---->
            </span>
            <!--v-if-->
          </span>
        </div>
        <!--v-if-->
        <portal-target multiple="true" name="app.files.sidebar.file.details.table" slot-props="[object Object]"></portal-target>
        <table aria-label="Overview of the information about the selected file" class="details-table">
          <tbody>
            <tr>
              <th class="oc-pr-s oc-font-semibold" scope="col">Last modified</th>
              <td>
                <span>Oct 21, 2015, 7:28 AM</span>
              </td>
            </tr>
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <tr>
              <th class="oc-pr-s oc-font-semibold" scope="col">Size</th>
              <td>24 kB</td>
            </tr>
            <!--v-if-->

            <tr>
              <th class="oc-pr-s oc-font-semibold" scope="col">Tags
                <!--v-if-->
              </th>
              <td>
                <div create-option="function createOption(label) {
      if (!label.trim().length) {
        return {
          label: label.toLowerCase().trim(),
          error: $gettext('Tag must not consist of blanks only'),
          selectable: false
        };
      }
      return {
        label: label.toLowerCase().trim()
      };
    }" map-keydown="function keydownMethods(map, vm) {
      var objectMapping = __assign({}, map);
      objectMapping[(0, keycode_1.default)('backspace')] = function (e) {
        return __awaiter(_this, void 0, void 0, function () {
          return __generator(this, function (_a) {
            switch (_a.label) {
              case 0:
                if (e.target.value || selectedTags.value.length === 0) {
                  return [2 /*return*/];
                }

                e.preventDefault();
                availableTags.value.push(selectedTags.value.pop());
                return [4 /*yield*/, save((0, vue_1.unref)(selectedTags))];
              case 1:
                _a.sent();
                return [2 /*return*/];
            }
          });
        });
      };

      return objectMapping;
    }" modelvalue="" options="" select-on-key-codes="13,188" selectable="function isOptionSelectable(option) {
      return (0, vue_1.unref)(selectedTags).length <= tagsMaxCount &amp;&amp; option.selectable !== false;
    }" taggable="">
                  <label class="oc-label" for="oc-select-1"></label>
                  <!--v-if-->
                  <div class="v-select vs--multiple vs--searchable oc-select" dir="auto" style="background: transparent;">
                    <div aria-expanded="false" aria-label="Search for option" aria-owns="vs1__listbox" class="vs__dropdown-toggle" id="vs1__combobox" role="combobox">
                      <div class="vs__selected-options">
                        <input aria-autocomplete="list" aria-controls="vs1__listbox" aria-labelledby="vs1__combobox" autocomplete="off" class="vs__search" id="oc-select-1" type="search">
                      </div>
                      <div class="vs__actions">
                        <button aria-label="Clear Selected" class="vs__clear" style="display: none;" title="Clear Selected" type="button">
                          <svg height="10" width="10" xmlns="http://www.w3.org/2000/svg">
                            <path d="M6.895455 5l2.842897-2.842898c.348864-.348863.348864-.914488 0-1.263636L9.106534.261648c-.348864-.348864-.914489-.348864-1.263636 0L5 3.104545 2.157102.261648c-.348863-.348864-.914488-.348864-1.263636 0L.261648.893466c-.348864.348864-.348864.914489 0 1.263636L3.104545 5 .261648 7.842898c-.348864.348863-.348864.914488 0 1.263636l.631818.631818c.348864.348864.914773.348864 1.263636 0L5 6.895455l2.842898 2.842897c.348863.348864.914772.348864 1.263636 0l.631818-.631818c.348864-.348864.348864-.914489 0-1.263636L6.895455 5z"></path>
                          </svg>
                        </button>
                        <svg class="vs__open-indicator" height="10" role="presentation" width="14" xmlns="http://www.w3.org/2000/svg">
                          <path d="M9.211364 7.59931l4.48338-4.867229c.407008-.441854.407008-1.158247 0-1.60046l-.73712-.80023c-.407008-.441854-1.066904-.441854-1.474243 0L7 5.198617 2.51662.33139c-.407008-.441853-1.066904-.441853-1.474243 0l-.737121.80023c-.407008.441854-.407008 1.158248 0 1.600461l4.48338 4.867228L7 10l2.211364-2.40069z"></path>
                        </svg>
                        <div class="vs__spinner" style="display: none;">Loading...</div>
                      </div>
                    </div>
                    <transition-stub appear="false" css="true" name="vs__fade" persisted="false">
                      <ul id="vs1__listbox" role="listbox" style="display: none; visibility: hidden;"></ul>
                    </transition-stub>
                  </div>
                  <!--v-if-->
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <ul class="oc-list oc-my-rm oc-mx-rm oc-mt-s" id="oc-files-actions-sidebar"></ul>
  </div>
</div>
`;
